rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp

```diff
@@ -89,23 +92,29 @@ class LifecycleNode::LifecycleNodeInterfaceImpl
     // The publisher takes a C-Typesupport since the publishing (i.e. creating
     // the message) is done fully in RCL.
     // Services are handled in C++, so that it needs a C++ typesupport structure.
-    rcl_ret_t ret = rcl_lifecycle_state_machine_init(
-      &state_machine_,
-      node_handle,
-      ROSIDL_GET_MSG_TYPE_SUPPORT(lifecycle_msgs, msg, TransitionEvent),
-      rosidl_typesupport_cpp::get_service_type_support_handle<ChangeStateSrv>(),
-      rosidl_typesupport_cpp::get_service_type_support_handle<GetStateSrv>(),
-      rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableStatesSrv>(),
-      rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableTransitionsSrv>(),
-      rosidl_typesupport_cpp::get_service_type_support_handle<GetAvailableTransitionsSrv>(),
-      &state_machine_options);
+    rcl_ret_t ret;
+    {
+      std::lock_guard<std::recursive_mutex> lock(state_machine_mutex_);
```

Comment:
Do we really need this scope for this lock?  It looks like we basically need to hold the lock until the end of the method, so I would suggest just taking the lock here and not dropping it until we leave the method.  That also means that re-taking the lock in `if (enable_communication_interface)` is no longer necessary.